luda-editor/migration/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+                    }
+
+                    #[inline]
+                    fn visit_some<D>(self, deserializer: D) -> Result<Value, D::Error>
+                    where
+                        D: serde::Deserializer<'de>,
+                    {
+                        Deserialize::deserialize(deserializer)
+                    }
+
+                    #[inline]
+                    fn visit_unit<E>(self) -> Result<Value, E> {
+                        Ok(Value::Null)
+                    }
+
+                    #[inline]
+                    fn visit_seq<V>(self, mut visitor: V) -> Result<Value, V::Error>
+                    where
+                        V: SeqAccess<'de>,
+                    {
+                        let mut vec = Vec::new();
+
+                        while let Some(elem) = tri!(visitor.next_element()) {
+                            vec.push(elem);
+                        }
+
+                        Ok(Value::Array(vec))
+                    }
+
+                    #[inline]
+                    fn visit_map<V>(self, mut visitor: V) -> Result<Value, V::Error>
+                    where
+                        V: MapAccess<'de>,
+                    {
+                        match visitor.next_key_seed(KeyClassifier)? {
+                            Some(KeyClass::Map(first_key)) => {
+                                let mut values = Map::new();
+
+                                values.insert(first_key, tri!(visitor.next_value()));
+                                while let Some((key, value)) = tri!(visitor.next_entry()) {
+                                    values.insert(key, value);
+                                }
+
+                                Ok(Value::Object(values))
+                            }
+                            None => Ok(Value::Object(Map::new())),
+                        }
+                    }
+                }
+
+                let mut json_value = deserializer.deserialize_any(ValueVisitor)?;
+                let json_object = json_value
+                    .as_object_mut()
+                    .ok_or_else(|| serde::de::Error::custom(format!("expected object")))?;
+
+                let version = json_object
+                    .get("_v")
+                    .map(|value| {
+                        value.as_u64().ok_or_else(|| {
+                            serde::de::Error::custom(format!("expected u64, but got {}", value))
+                        })
+                    })
+                    .unwrap_or(Ok(0))?;
+
+                if version < 1 {
+                    let previous = serde_json::from_value(json_value).map_err(|e| {
+                        serde::de::Error::custom(format!("failed to deserialize previous version: {}", e))
+                    })?;
+                    Ok(#struct_name::migrate(previous))
+                } else {
+                    Ok(#struct_name {
+                        #(#fields_deserialize)*
+                    })
+                }
```

Comment:
Could you check this?
Is it right to use `#struct_name::migrate(previous)` when serialized version is lower than **1**?. not **current version**?
I expected the following logic.
```
if `serialized_version` is lower than `current_version`:
    do migrate
else:
    return deserialized
```

Reply:
Oh my typo. give me sec

Reply:
Fixed https://github.com/NamseEnt/namseent/pull/412/commits/8e3cb246ed1cf1f8652a42272e56aed93400d255